bitbake/lib/bb/fetch2/npm.py

```diff
@@ -2,300 +2,200 @@
-        ud.pkgdatadir = d.expand("${DL_DIR}/%s" % prefixdir)
-        if not os.path.exists(ud.pkgdatadir):
-            bb.utils.mkdirhier(ud.pkgdatadir)
-        ud.localpath = d.expand("${DL_DIR}/npm/%s" % ud.bbnpmmanifest)
-
-        self.basecmd = d.getVar("FETCHCMD_wget") or "/usr/bin/env wget -O -t 2 -T 30 -nv --passive-ftp --no-check-certificate "
-        ud.prefixdir = prefixdir
-
-        ud.write_tarballs = ((d.getVar("BB_GENERATE_MIRROR_TARBALLS") or "0") != "0")
-        mirrortarball = 'npm_%s-%s.tar.xz' % (ud.pkgname, ud.version)
-        mirrortarball = mirrortarball.replace('/', '-')
-        ud.fullmirror = os.path.join(d.getVar("DL_DIR"), mirrortarball)
-        ud.mirrortarballs = [mirrortarball]
+            raise ParameterError("Parameter 'version' required", ud.url)
+
+        # Get the 'registry' part of the url
+        ud.registry = ud.url.replace("npm://", "http://", 1).split(";")[0]
+
+        # Using the 'downloadfilename' parameter as local filename or the
+        # npm package name.
+        if "downloadfilename" in ud.parm:
+            ud.basename = ud.parm["downloadfilename"]
+        else:
+            # Scoped package names (with the @) use the same naming convention
+            # as the 'npm pack' command.
+            if ud.name.startswith("@"):
+                ud.basename = re.sub("/", "-", ud.name[1:])
+            else:
+                ud.basename = ud.name
+            ud.basename += "-" + ud.version + ".tgz"
+
+        ud.localfile = d.expand(ud.basename)
+
+        ud.basecmd = d.getVar("FETCHCMD_wget")
+
+        if not ud.basecmd:
+            ud.basecmd = "wget"
+            ud.basecmd += " --tries=2"
+            ud.basecmd += " --timeout=30"
+            ud.basecmd += " --passive-ftp"
+            ud.basecmd += " --no-check-certificate"
+
+    @staticmethod
+    def _run_npm_view(ud, d):
+        """
+            Run the 'npm view' command to get informations about a npm package.
+        """
+
+        cmd = "npm view"
+        cmd += " --json"
+        cmd += " --registry {}".format(ud.registry)
+        cmd += " '{}@{}'".format(ud.name, ud.version)
+
+        view_string = runfetchcmd(cmd, d)
+
+        view = json.loads(view_string)
+
+        if isinstance(view, list):
+            return view[-1]
+
+        return view
 
     def need_update(self, ud, d):
+        """
+            Force a fetch, even if localpath exists?
+        """
+
+        # Note that when using a version which does not exist on the registry
+        # (like 'version=latest') the local filename will use this version
+        # string (e.g 'my-package-latest.tgz') but the actual downloaded file
+        # will be the latest version available on the registry (e.g '1.2.3').
+        #
+        # This trick can be useful with devtool / recipetool to automatically
+        # fetch the latest version, BUT these files (the ones with non-existent
+        # versions) must be updated every times.
```

Comment:
If I understand well, it is not  related to 'latest' keyword it could be anything else ?
If its the case it is dangerous, the described behavior should only be applied in case of 'latest' version specified.

Reply:
Your right. I will update the behavior to accept only existent version or the "latest" keyword.